bpf/bpf_lxc.c

```diff
@@ -1258,6 +1262,10 @@ static __always_inline int __tail_handle_ipv4(struct __ctx_buff *ctx)
 				goto skip_service_lookup;
 			}
 #endif /* ENABLE_L7_LB */
+
+			if (is_v4_loopback(ip4->daddr))
+				goto skip_service_lookup;
+
```

Comment:
I think this is not enough given also traffic on the wire could get to the node where we have the 127.0.0.1 as IP in svc table, so also this needs to be rejected from devs with bpf_host/xdp on it.

I wonder.. are hostport ranges forbidden to clash with nodeport ranges? If so, maybe we could solve this issue by installing a wildcard entry, which covers loopback and all node local addresses - similar as we do in kpr for nodeport.

The logic in sock LB, should probably go into sock{4,6}_skip_xlate() which already has checks around lb{4,6}_svc_is_hostport().

Reply:
Thanks @borkmann !

Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

For the port range: I believe that the hostPort range is not defined, so it can be any, but the NodePort range is reserved, so we won't let users to create HostPorts within the NodePort range. As a matter of fact, I've just checked that you added this check in 2020 in a commit named "prep datapath for loopback/local address exposure on hostport", so I will look into installing the wildcard entry :-)

Reply:
> Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

Yes, this exactly, e.g. you could craft af_packet skb and push it to the wire when in same L2.